crates/core/src/host/wasmer/mod.rs

```diff
@@ -62,52 +67,90 @@ pub fn make_actor(
 }
 
 impl Mem {
     /// Constructs an instance of `Mem` from an exports map.
-    fn extract(exports: &wasmer::Exports) -> anyhow::Result<Self> {
-        let memory = exports.get_memory("memory")?.clone();
-        Ok(Self { memory })
+    fn extract(exports: &wasmtime::Instance, store: impl wasmtime::AsContextMut) -> anyhow::Result<Self> {
+        Ok(Self {
+            memory: exports.get_memory(store, "memory").context("no memory export")?,
+        })
     }
 
     /// Creates and returns a view into the actual memory `store`.
     /// This view allows for reads and writes.
-
-    fn view<'a>(&self, store: &'a impl AsStoreRef) -> wasmer::MemoryView<'a> {
-        self.memory.view(store)
+    fn view_and_store<'a, T>(&self, store: impl Into<StoreContextMut<'a, T>>) -> (&'a mut MemView, &'a mut T) {
+        let (mem, store_data) = self.memory.data_and_store_mut(store);
+        (MemView::from_slice_mut(mem), store_data)
     }
 
-    /// Reads a slice of bytes starting from `ptr`
-    /// and lasting `len` bytes into a `Vec<u8>`.
-    ///
-    /// Returns an error if the slice length overflows a 64-bit address.
-    fn read_bytes(&self, store: &impl AsStoreRef, ptr: WasmPtr<u8>, len: u32) -> Result<Vec<u8>, MemoryAccessError> {
-        ptr.slice(&self.view(store), len)?.read_to_vec()
+    fn view_const<'a, T: 'a>(&self, store: impl Into<StoreContext<'a, T>>) -> &'a MemView {
+        MemView::from_slice(self.memory.data(store))
     }
+}
+
+#[repr(transparent)]
+struct MemView([u8]);
 
-    /// Writes `data` into the slice starting from `ptr`
-    /// and lasting `len` bytes.
-    ///
-    /// Returns an error if
-    /// - the slice length overflows a 64-bit address
-    /// - `len != data.len()`
-    fn set_bytes(
-        &self,
-        store: &impl AsStoreRef,
-        ptr: WasmPtr<u8>,
-        len: u32,
-        data: &[u8],
-    ) -> Result<(), MemoryAccessError> {
-        ptr.slice(&self.view(store), len)?.write_slice(data)
+impl MemView {
+    fn from_slice_mut(v: &mut [u8]) -> &mut Self {
+        unsafe { &mut *(v as *mut [u8] as *mut MemView) }
+    }
+    fn from_slice(v: &[u8]) -> &Self {
+        unsafe { &*(v as *const [u8] as *const MemView) }
+    }
+    fn deref_slice(&self, offset: WasmPtr<u8>, len: u32) -> Result<&[u8], WasmError> {
+        self.0
+            .get(offset as usize..)
+            .and_then(|s| s.get(..len as usize))
+            .ok_or_else(|| WasmError::Wasm(wasmtime::Trap::MemoryOutOfBounds.into()))
+    }
+    fn deref_str(&self, offset: WasmPtr<u8>, len: u32) -> Result<&str, WasmError> {
+        let b = self.deref_slice(offset, len)?;
+        std::str::from_utf8(b).map_err(|e| WasmError::Wasm(e.into()))
+    }
+    fn deref_str_lossy(&self, offset: WasmPtr<u8>, len: u32) -> Result<Cow<str>, WasmError> {
+        self.deref_slice(offset, len).map(String::from_utf8_lossy)
+    }
+    fn opt_deref_str_lossy(&self, offset: WasmPtr<u8>, len: u32) -> Result<Option<Cow<str>>, WasmError> {
```

Comment:
```suggestion
    /// Convert `self[offset..offset+len]` to a string,
    /// replacing invalid UTF-8 bytes with U+FFFD REPLACEMENT CHARACTER.
    /// Returns `None` if `offset` is 0, the null pointer.
    fn try_deref_str_lossy(&self, offset: WasmPtr<u8>, len: u32) -> Result<Option<Cow<str>>, WasmError> {
```

Reply:
Hmmm, idk about the try naming - I'm not a huge fan of `opt_` as a prefix either, but try implies the deref is a fallible operation, when it's really just treating null as a sentinel value. I could do a follow up fixing this, maybe it could be one set of functions without the `opt_` prefix but that return `Option<>`, and then a extension trait with a `fn non_nullable(self: Result<Option<T>, WasmError>) -> Result<T, WasmError>;` so that the calls look like `mem.deref_str(buf, buf_len).non_nullable()?`

Reply:
Oh, or I guess I could do it here. I forgot that I added those functions in this branch, lol